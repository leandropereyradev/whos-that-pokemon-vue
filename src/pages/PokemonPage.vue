<template>
  <h1 v-if="!pokemon">Loading...</h1>

  <div v-else>
    <h1>Who's That Pokemon??</h1>

    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />

    <PokemonOptions :pokemons="pokemonsArr" @selected-pokemon="checkAnswer($event)" />

    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
    </template>
  </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue';
import PokemonOptions from '@/components/PokemonOptions.vue';

import getPokemonsOptions from '@/helpers/getPokemonOptions'

export default {
  name: 'Pokemon Page',
  components: {
    PokemonPicture,
    PokemonOptions
  },
  data() {
    return {
      pokemonsArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ""
    }
  },
  methods: {
    async mixPokemonsArray() {
      this.pokemonsArr = await getPokemonsOptions()

      const randomIndexSelectedPokemons = Math.floor(Math.random() * 4)

      this.pokemon = this.pokemonsArr[randomIndexSelectedPokemons]
    },
    checkAnswer(selectedPokemonId) {
      this.showPokemon = true;
      this.showAnswer = true

      if (this.pokemon.id === selectedPokemonId) {
        this.message = `Great!!! It's ${this.pokemon.name.charAt(0).toUpperCase() + this.pokemon.name.slice(1)}`
      } else {
        this.message = `Oops!!! This Pokemon is ${this.pokemon.name.charAt(0).toUpperCase() + this.pokemon.name.slice(1)}`
      }
    }
  },
  mounted() {
    this.mixPokemonsArray()
  }
}
</script>