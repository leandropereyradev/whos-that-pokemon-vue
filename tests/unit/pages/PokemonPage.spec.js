import { shallowMount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage";
import { mockPokemons } from "../mocks/pokemons.mock";

describe("PokemonPage component", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });

  test("Should call the mixPokemonsArray when mounting", () => {
    const mixPokemonsArraySpy = jest.spyOn(
      PokemonPage.methods,
      "mixPokemonsArray"
    );

    shallowMount(PokemonPage);

    expect(mixPokemonsArraySpy).toHaveBeenCalled();
  });

  test("Should match this snapshot when it loads the Pokemons", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonsArr: mockPokemons,
          pokemon: mockPokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
          isDisabled: false,
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test("Should show the PokemonPicture and PokemonOptions components", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonsArr: mockPokemons,
          pokemon: mockPokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
          isDisabled: false,
        };
      },
    });

    const pictureTag = wrapper.find("pokemon-picture-stub");
    const optionsTag = wrapper.find("pokemon-options-stub");

    expect(pictureTag.exists()).toBeTruthy();
    expect(optionsTag.exists()).toBeTruthy();

    expect(pictureTag.attributes("pokemonid")).toBe("1");
    expect(optionsTag.attributes("pokemons")).toBeTruthy();
  });

  test("The checkAnswer() function should be executed", async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonsArr: mockPokemons,
          pokemon: mockPokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
          isDisabled: false,
        };
      },
    });

    const pokemonName =
      mockPokemons[0].name.charAt(0).toUpperCase() +
      mockPokemons[0].name.slice(1);

    await wrapper.vm.checkAnswer(1);

    expect(wrapper.find("h2").exists()).toBeTruthy();
    expect(wrapper.vm.showPokemon).toBeTruthy();
    expect(wrapper.find("h2").text()).toBe(`Great!!! It's ${pokemonName}`);

    await wrapper.vm.checkAnswer(2);
    expect(wrapper.vm.message).toBe(`Oops!!! This is ${pokemonName}`);
  });
});
