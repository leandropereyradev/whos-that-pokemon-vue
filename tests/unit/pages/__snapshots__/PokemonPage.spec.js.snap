// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component Should match this snapshot when it loads the Pokemons 1`] = `
<div class="container">
  <h1>Who's That Pokemon??</h1>
  <div class="pokedex"><button class="btnReset btn animation">Reset</button>
    <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
    <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]" isdisabled="false"></pokemon-options-stub>
  </div>
  <!--v-if-->
</div>
`;
